File: py/01-hashing/022-intersection-with-dupes.py
def intersection_plus(a: list, b: list) -> list:
    # idea 1:
    # dict for each, compare occurrences, use the min value to append as required

    group_a = {}
    group_b = {}

    for item in a:
        if item in group_a:
            group_a[item] += 1
        else:
            group_a[item] = 1

    for item in b:
        if item in group_b:
            group_b[item] += 1
        else:
            group_b[item] = 1

    intersection = []

    for item in group_a:
        if item in group_b:
            intersection.extend([item]*min(group_a[item], group_b[item]))

    return intersection


# test cases
print(intersection_plus(
  ["a", "b", "c", "b"], 
  ["x", "y", "b", "b"]
)) # -> ["b", "b"]
print(intersection_plus(
  ["q", "b", "m", "s", "s", "s"], 
  ["s", "m", "s"]
)) # -> ["m", "s", "s"]
print(intersection_plus(
  ["p", "r", "r", "r"], 
  ["r"]
)) # -> ["r"]
print(intersection_plus(
  ["r"], 
  ["p", "r", "r", "r"]
)) # -> ["r"]
print(intersection_plus(
  ["t", "v", "u"], 
  ["g", "e", "d", "f"]
)) # -> [ ]
print(intersection_plus(
  ["a", "a", "a", "a", "a", "a",], 
  ["a", "a", "a", "a"]
)) # -> ["a", "a", "a", "a"]

File: js/01-hashing/022-intersection-with-dupes.js
function intersectionPlus(a, b) {
    let hash_a = {};
    let hash_b = {};
    let result = [];
    for (let i = 0; i < a.length; i++) {
        if (!hash_a[i]) {
            hash_a[i] = 1;
        } else {
            hash_a[i]++;
        };
    };

    for (let i = 0; i < b.length; i++) {
        if (!hash_b[i]) {
            hash_b[i] = 1;
        } else {
            hash_b[i]++;
        };
    };

    for (key in hash_a) {
        if (hash_b[hash_a[key]]) {
            let i=0;
            while (i<=Math.min(hash_a[key], hash_b[key])) {
                result.push(hash_b[key]);
                i++;
            };
        };
    };

    return result;
};

// test cases
console.log(intersectionPlus(
    ["a", "b", "c", "b"],
    ["x", "y", "b", "b"]
)) // -> ["b", "b"]
console.log(intersectionPlus(
    ["q", "b", "m", "s", "s", "s"],
    ["s", "m", "s"]
)) // -> ["m", "s", "s"]
console.log(intersectionPlus(
    ["p", "r", "r", "r"],
    ["r"]
)) // -> ["r"]
console.log(intersectionPlus(
    ["r"],
    ["p", "r", "r", "r"]
)) // -> ["r"]
console.log(intersectionPlus(
    ["t", "v", "u"],
    ["g", "e", "d", "f"]
)) // -> []
console.log(intersectionPlus(
    ["a", "a", "a", "a", "a", "a",],
    ["a", "a", "a", "a"]
)) // -> ["a", "a", "a", "a"]